PHPCompatibility/Sniffs/Syntax/NewConstantDereferencingSniff.php

```diff
@@ -0,0 +1,135 @@
+<?php
+/**
+ * PHPCompatibility, an external standard for PHP_CodeSniffer.
+ *
+ * @package   PHPCompatibility
+ * @copyright 2012-2021 PHPCompatibility Contributors
+ * @license   https://opensource.org/licenses/LGPL-3.0 LGPL3
+ * @link      https://github.com/PHPCompatibility/PHPCompatibility
+ */
+
+namespace PHPCompatibility\Sniffs\Syntax;
+
+use PHPCompatibility\Sniff;
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Util\Tokens;
+
+/**
+ * Detect constant dereferencing.
+ *
+ * As of PHP 5.6, constants can now be dereferenced to access individual elements and characters.
+ *
+ * As of PHP 8.0, constants can also be dereferenced to invoke method calls, though this support only applies to syntax and must be used in conjunction with the scalar objects extension to work at runtime.
+ *
+ * PHP version 5.6
+ * PHP version 8.0
+ *
+ * @link https://wiki.php.net/rfc/const_scalar_exprs
+ * @link https://github.com/php/php-src/commit/d5ddd2dbb263cd626a5e0f203c00d6f39168507b
+ * @link https://wiki.php.net/rfc/variable_syntax_tweaks#constant_dereferencability
+ *
+ * @since 10.0.0
+ */
+class NewConstantDereferencingSniff extends Sniff
+{
+
+    /**
+     * Returns an array of tokens this test wants to listen for.
+     *
+     * @since 10.0.0
+     *
+     * @return array
+     */
+    public function register()
+    {
+        return [
+            T_STRING,
+        ];
+    }
+
+    /**
+     * Processes this test, when one of its tokens is encountered.
+     *
+     * @since 10.0.0
+     *
+     * @param \PHP_CodeSniffer\Files\File $phpcsFile The file being scanned.
+     * @param int                         $stackPtr  The position of the current token in
+     *                                               the stack passed in $tokens.
+     *
+     * @return void
+     */
+    public function process(File $phpcsFile, $stackPtr)
+    {
+        // Impacted features are supported in PHP 8, bail early
+        if ($this->supportsBelow('7.4') === false) {
+            return;
+        }
+
+        $tokens = $phpcsFile->getTokens();
+
+        // Not a reference, out of scope
+        $referenceTokens  = [
```

Comment:
Can this array be turned into a class property ? and as a property possibly given a value which would be helpful to do the check on line 114 and 126 ? (i.e. which error should be thrown for what)

Reply:
The array here contains `T_OBJECT_OPERATOR` in addition to the other two tokens also used on 126. That would leave us doing something like this here:

```php
if (!isset($this->referenceTokens[$nextNonEmptyCode])
    && $nextNonEmptyCode !== \T_OBJECT_OPERATOR) {
```

Is that what you had in mind?